File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
sys.setrecursionlimit(10000)
read = sys.stdin.readline

N, M, V = map(int, read().split())
L = [[] for _ in range(N + 1)]

def DFS(num):
    visited[num] = True
    print(num, end = ' ')
    for i in L[num]:
        if not visited[i]:
            DFS(i)

def BFS(num):
    q = deque()
    q.append(num)
    visited[num] = True

    while q:
        i = q.popleft()
        print(i, end = ' ')
        for j in L[i]:
            if not visited[j]:
                q.append(j)
                visited[j] = True

for i in range(M):
    a, b = map(int, read().split())
    L[a].append(b)
    L[b].append(a)

for i in range(1, N + 1):
    L[i].sort()

visited = [False] * (N + 1)
DFS(V)
print()
visited = [False] * (N + 1)
BFS(V)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static List<Integer>[] graph;
    public static boolean[] visited;
    public static StringBuilder sb;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int V = Integer.parseInt(st.nextToken());

        graph = new List[N + 1];
        for (int i = 1; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph[a].add(b);
            graph[b].add(a);
        }

        for (int i = 1; i <= N; i++) {
            Collections.sort(graph[i]);
        }

        sb = new StringBuilder();

        visited = new boolean[N + 1];
        dfs(V);
        sb.append("\n");

        visited = new boolean[N + 1];
        bfs(V);
        sb.append("\n");

        System.out.println(sb);
    }

    public static void dfs (int num) {
        visited[num] = true;
        sb.append(num).append(" ");

        for (int next : graph[num]) {
            if (!visited[next]) {
                dfs(next);
            }
        }
    }

    public static void bfs (int num) {
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(num);

        visited[num] = true;

        while (!q.isEmpty()) {
            int cur = q.poll();

            sb.append(cur).append(" ");

            for (int next : graph[cur]) {
                if (!visited[next]) {
                    visited[next] = true;
                    q.offer(next);
                }
            }
        }
    }
}
